app/src/main/java/app/grapheneos/camera/ui/composable/screen/GalleryScreen.kt

```diff
@@ -0,0 +1,349 @@
+import androidx.compose.ui.platform.LocalContext
+import androidx.compose.ui.unit.LayoutDirection
+import androidx.compose.ui.unit.dp
+import app.grapheneos.camera.CapturedItem
+import app.grapheneos.camera.R
+import app.grapheneos.camera.ktx.showOrReplaceSnackbar
+import app.grapheneos.camera.ui.activities.VideoPlayer
+import app.grapheneos.camera.ui.composable.components.FileDeletionDialog
+import app.grapheneos.camera.ui.composable.components.GalleryImage
+import app.grapheneos.camera.ui.composable.components.GalleryTopBar
+import app.grapheneos.camera.ui.composable.components.MediaInfoDialog
+import app.grapheneos.camera.ui.theme.AppColor
+import kotlinx.coroutines.launch
+import me.saket.telephoto.zoomable.rememberZoomableState
+import me.saket.telephoto.zoomable.zoomable
+
+import app.grapheneos.camera.ITEM_TYPE_IMAGE
+import app.grapheneos.camera.ITEM_TYPE_VIDEO
+import app.grapheneos.camera.ui.composable.data.MediaItemDetails
+
+private const val TAG = "GalleryScreen"
+
+@OptIn(ExperimentalFoundationApi::class)
+@Composable
+fun GalleryScreen(
+    isSecureMode: Boolean,
+    mediaItemsLoader: () -> List<CapturedItem>,
+    latestMediaItemLoader: () -> CapturedItem?,
+    onExit: () -> Unit = {},
+) {
+
+    val context = LocalContext.current
+
+    val coroutineScope = rememberCoroutineScope()
+
+    val snackbarHostState = remember { SnackbarHostState() }
+
+    val zoomableState = rememberZoomableState()
+
+    var isZoomedIn by remember {
+        mutableStateOf(false)
+    }
+
+    var inFocusMode by remember {
+        mutableStateOf(false)
+    }
+
+    val backgroundColor by animateColorAsState(
+        label = "background_color_animation",
+        targetValue = if (inFocusMode) Color.Black else AppColor.BackgroundColor,
+        animationSpec = tween(durationMillis = 300, easing = EaseIn),
+    )
+
+    val capturedItems = remember {
+        mutableStateListOf<CapturedItem>()
+    }
+
+    val pagerState = rememberPagerState(pageCount = {
+        capturedItems.size
+    })
+
+    val displayedMediaInfo = rememberSaveable(saver = MediaItemDetails.Saver) {
+        mutableStateOf(null)
+    }
+
+    var itemToBeDeleted by rememberSaveable {
+        mutableStateOf<CapturedItem?>(null)
+    }
+
+    fun getCurrentMediaItem(): CapturedItem {
+        return capturedItems[pagerState.currentPage]
+    }
+
+    LaunchedEffect(mediaItemsLoader) {
+        coroutineScope.launch {
+            val latestItem = latestMediaItemLoader()
+            synchronized(capturedItems) {
+                if (latestItem != null && capturedItems.isEmpty()) {
+                    capturedItems.add(latestItem)
+                }
+            }
+        }
+
+        coroutineScope.launch {
+            val items = mediaItemsLoader()
+
+            if (items.isEmpty()) {
+                Toast.makeText(context, R.string.empty_gallery, Toast.LENGTH_LONG).show()
+                onExit()
+            }
+
+            capturedItems.clear()
```

Comment:
Missing synchronization, and ordering between these 2 coroutines doesn't look right. There's no guarantee about which will finish first.

Reply:
We have added checks to avoid unexpected behavior irrespective of which finishes first (`latestItem != null && capturedItems.isEmpty()`). We could merge it into a single coroutine callback to keep it simple and avoid the need for synchronization altogether